Devfile Template Guide (#39)
devfile/devfile-web: libs/docs/src/docs/2.2.0/create-devfiles-with-templates.md

---

> title: Creating Devfiles With Templates

Reviewer: ```suggestion
title: Creating Devfiles With templates
```

description: Creating devfiles with templates
---

Most dev tools can utilize the devfile registry to 
fetch the stacks needed to begin development on projects. 
Sometimes, however, you might need to create a devfile 
from scratch or to create your own devfile template(s). 
This guide runs through the process, starting from a minimum devfile and 
building sample templates for common use cases.

## Prerequisites

- [Versions](./versions)

> - [Metadata](./metadata)

Reviewer 2: Will these links work? These topics are under the `/docs/no-version` folder not the current version specific location

Author: @kim-tsao Good point, though it looks to work as it seems no-version content get placed under all versions on build.

- [Adding components](./adding-components)
- [Adding commands](./adding-commands)

> 

Reviewer 2: Since the procedure goes through adding a starter project, we should also add a link to the topic [Adding projects](./adding-projects)

Author: Added [Defining starter projects](https://devfile.io/docs/2.2.0/defining-starter-projects).

## Procedure

> 1. Creating a minimal devfile

Reviewer: ```suggestion
1. Creating a minimal devfile:
```

    - `schemaVersion` is the only required root element
    - `metadata` is optional but it is recommended to have in your 
    templates
    ```yaml {% title="Minimal Devfile" filename="devfile.yaml" %}
    schemaVersion: 2.2.0
    ```
    ```yaml {% title="Minimal Devfile with Metadata" filename="devfile.yaml" %}
    schemaVersion: 2.2.0
    metadata:
      name: devfile-sample
      version: 2.0.0
    ```

> 2. Creating a web service template

Reviewer: ```suggestion
2. Creating a web service template:
```

    1. Template `metadata`
        - Set the template `name` and `version`
        - Set a `description` for the template
        - Set the `projectType` and `language` to describe what
        kind of stack is being used
        - Set `provider` to tell who is providing this template
        - Set `tags` to give keyword which describe the elements
        of the stack
        - Set `architectures` to specify the platforms support in 
        the template
        - For improved readability on devfile registries, set `displayName` to the title that will be the display text for this template and `icon` to tie a stack icon to the template:
        ```yaml
          schemaVersion: 2.2.0
          metadata:
            name: web-service
            version: 1.0.0
            description: A web service template.
            projectType: Go
            language: Go
            provider: Red Hat
            tags: [ 'Go', 'Gin', 'pq' ]
            architectures: [ 'amd64' ]
            displayName: Simple Web Service
            icon: https://raw.githubusercontent.com/devfile-samples/devfile-stack-icons/main/golang.svg
          ```
    2. Setup `components`

>         - A `name` is required for a component.

Reviewer: I would go with a single style for list items, when they are complete sentences - the first part of the list does not use periods at the end of each item, while the second one does.

Author: @gabriel-rh I changed these instances to use sub-lists rather than multiple sentences and removed ending periods on all list items.

        - If a `container` entity is defined, a `image` must be 
        specified
        - Though `endpoints` is optional it is needed to expose 
        the port for web connections, this requires an endpoint `name`

>         and `targetPort` to expose (e.g. `8080` if that is your http 

Reviewer: ```suggestion
        and `targetPort` to expose (for example, `8080` if that is your http 
```

>         and `targetPort` to expose, for example `8080` if that is your http 

Reviewer: ```suggestion
        and `targetPort` to expose, for example, `8080` if that is your http 
```

        port.

>         - The web service might need to connect to a external 

Reviewer: ```suggestion
        - The web service might need to connect to an external 
```

>         database using environment variables, in this case 

Reviewer: ```suggestion
        database using environment variables. In this case, define
```

>         environment variable names and values defined for the 

Reviewer: ```suggestion
        environment variable names and values for the 
```

        container component under `env`
        ```yaml
        components:
          - name: web
            container:
              endpoints:
                - name: http
                  targetPort: 8080
              env:
                DATABASE_HOST: db.example.com
                DATABASE_PORT: 5432
                DATABASE_NAME: dev
                DATABASE_USER: devuser
                DATABASE_PASSWORD: devpassword
              image: quay.io/devfile/golang:latest
        ```
    3. Adding `commands`
        - An `id` to identify the command
        - An `exec` entity must be defined with a `commandLine` string
        and a reference to a `component`. This implies that at least 
        one component entity is defined under the root `components`
        element.

>         - Set the `workingDir` to where the project source is 

Reviewer: ```suggestion
        - The `workingDir` is set to where the project source is 
```

        stored

>         - Command groups can be used to define automation useful

Reviewer: ```suggestion
        - Command groups can be used to define automation that is useful
```

        for executing a web service
        ```yaml
        commands:
          - id: build
            exec:
              commandLine: go build main.go
              component: web
              workingDir: ${PROJECT_SOURCE}
              group:
                kind: build
                isDefault: true
          - id: run
            exec:
              commandLine: ./main
              component: web
              workingDir: ${PROJECT_SOURCE}
              group:
                kind: run
                isDefault: true
        ```
    4. Define starter projects
        - Add a starter project under `starterProjects` which

>         includes at least a `name` and remote location, either

Reviewer: ```suggestion
        including at least a `name` and remote location, either
```

        `git` or `zip`
        - It is recommended to include a starter project 

>         description

Reviewer: ```suggestion
        description:
```

          ```yaml
          starterProjects:
            - name: web-starter
              description: A web service starting point.
              git:
                remotes:
                  origin: https://github.com/devfile-samples/devfile-stack-go.git
          ```

>     ```yaml {% title="Complete Web Service Template" filename="devfile.yaml" %}

Reviewer 2: Should we have a title before this line describing the output? e.g. `Putting it all together, this is what the complete devfile looks like`
